### PR TITLE
feat: add metawasm state impl

### PR DESCRIPTION
**Pull request by bezrazli4n0**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,39 +1,44 @@
...
+	@echo ⚙️ Installing a toolchain \& a target...
 	@rustup toolchain add nightly
 	@rustup target add wasm32-unknown-unknown --toolchain nightly
 
-linter:
-	@echo ──────────── Run linter ───────────────────────
-	@cargo +nightly clippy --all-targets -- --no-deps -D warnings -A "clippy::missing_safety_doc"
+lint:
+	@echo ⚙️ Running the linter...
+	@cargo +nightly clippy --all-targets -- -D warnings
```
Comment (fluiderson) [inline Makefile:25]:
```suggestion
	@cargo +nightly clippy --workspace --all-targets -- -D warnings
```
To run the linter on the `state` crate too.

**Review by fluiderson — COMMENTED** [review]
Please fix CI

File: src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+            Some(member) => {
+                if member.shares == 0 {
+                    return false;
+                }
+            }
+            None => {
+                return false;
+            }
+        }
+        true
```
Comment (shamilsan) [inline src/contract.rs:345]: Might be simpler
```suggestion
        matches!(self.members.get(account), Some(member) if member.shares > 0)
```

File: src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+    // check that `msg::source()` is either a DAO member or a delegate key
+    fn check_for_membership(&self) {
+        if !self.is_member(&msg::source()) {
+            panic!("account is not a DAO member")
+        }
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {
```
Comment (shamilsan) [inline src/contract.rs:357]: Please no `pub` and `unsafe`
```suggestion
extern "C" fn init() {
```

File: src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    let config: InitDao = msg::load().expect("Unable to decode InitDao");
+    let dao = Dao {
+        approved_token_program_id: config.approved_token_program_id,
+        voting_period_length: config.voting_period_length,
+        period_duration: config.period_duration,
+        ..Dao::default()
+    };
+    DAO = Some(dao);
```
Comment (shamilsan) [inline src/contract.rs:365]:
```suggestion
    unsafe { DAO = Some(dao) };
```

File: src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+}
+
+#[gstd::async_main]
+async unsafe fn main() {
```
Comment (shamilsan) [inline src/contract.rs:369]: Also
```suggestion
async fn main() {
```

File: src/contract.rs [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+            Vote::Yes => {
+                proposal.yes_votes = proposal.yes_votes.saturating_add(member.shares);
+                // it is necessary to save the highest id of the proposal - must be processed for member to ragequit
+                if let Some(id) = member.highest_index_yes_vote {
+                    if id < proposal_id {
+                        member.highest_index_yes_vote = Some(proposal_id);
+                    }
+                } else {
+                    member.highest_index_yes_vote = Some(proposal_id);
+                }
```
Comment (shamilsan) [inline src/contract.rs:195]: Might be shorted and clearer:
```suggestion
                let id = member.highest_index_yes_vote.get_or_insert(proposal_id);
                *id = proposal_id.max(*id);
```

File: io/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+#[derive(Debug, Decode, Encode, TypeInfo)]
+pub struct InitDao {
+    pub approved_token_program_id: ActorId,
+    pub voting_period_length: u64,
+    pub period_duration: u64,
+    pub grace_period_length: u64,
+}
+
+#[derive(Debug, Encode, Decode, Clone, TypeInfo)]
+pub enum Vote {
```
Comment (shamilsan) [inline io/src/lib.rs:127]: Such lightweight enum might be `Copy` to simplify the code.

**Review by shamilsan — APPROVED** [review]

**Review by LouiseMedova — APPROVED** [review]

**Pull request merged**